File: dxbase/src/main/java/com/dxmovie/dxbase/binding/viewadapter/listview/ViewAdapter.java
package com.dxmovie.dxbase.binding.viewadapter.listview;

import androidx.databinding.BindingAdapter;
import android.view.View;
import android.widget.AbsListView;
import android.widget.AdapterView;
import android.widget.ListView;

import java.util.concurrent.TimeUnit;

import io.reactivex.functions.Consumer;
import io.reactivex.subjects.PublishSubject;
import com.dxmovie.dxbase.binding.command.BindingCommand;
import com.dxmovie.dxbase.binding.command.BindingCommandWithParams;

/**
 */
public final class ViewAdapter {

    @SuppressWarnings("unchecked")
    @BindingAdapter(value = {"onScrollChangeCommand", "onScrollStateChangedCommand"}, requireAll = false)
    public static void onScrollChangeCommand(final ListView listView,
                                             final BindingCommandWithParams<ListViewScrollDataWrapper> onScrollChangeCommand,
                                             final BindingCommandWithParams<Integer> onScrollStateChangedCommand) {
        listView.setOnScrollListener(new AbsListView.OnScrollListener() {
            private int scrollState;

            @Override
            public void onScrollStateChanged(AbsListView view, int scrollState) {
                this.scrollState = scrollState;
                if (onScrollStateChangedCommand != null) {
                    onScrollStateChangedCommand.executeAction(scrollState);
                }
            }

            @Override
            public void onScroll(AbsListView view, int firstVisibleItem, int visibleItemCount, int totalItemCount) {
                if (onScrollChangeCommand != null) {
                    onScrollChangeCommand.executeAction(new ListViewScrollDataWrapper(scrollState, firstVisibleItem, visibleItemCount, totalItemCount));
                }
            }
        });

    }


    @BindingAdapter(value = {"onItemClickCommand"}, requireAll = false)
    public static void onItemClickCommand(final ListView listView, final BindingCommandWithParams<Integer> onItemClickCommand) {
        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                if (onItemClickCommand != null) {
                    onItemClickCommand.executeAction(position);
                }
            }
        });
    }


    @BindingAdapter({"onLoadMoreCommand"})
    public static void onLoadMoreCommand(final ListView listView, final BindingCommandWithParams<Integer> onLoadMoreCommand) {
        listView.setOnScrollListener(new OnScrollListener(listView, onLoadMoreCommand));

    }

    public static class OnScrollListener implements AbsListView.OnScrollListener {
        private PublishSubject<Integer> methodInvoke = PublishSubject.create();
        private BindingCommandWithParams<Integer> onLoadMoreCommand;
        private ListView listView;

        public OnScrollListener(ListView listView, final BindingCommandWithParams<Integer> onLoadMoreCommand) {
            this.onLoadMoreCommand = onLoadMoreCommand;
            this.listView = listView;
            methodInvoke.throttleFirst(1, TimeUnit.SECONDS)
                    .subscribe(new Consumer<Integer>() {
                        @Override
                        public void accept(Integer integer) throws Exception {
                            onLoadMoreCommand.executeAction(integer);
                        }
                    });
        }

        @Override
        public void onScrollStateChanged(AbsListView view, int scrollState) {

        }

        @Override
        public void onScroll(AbsListView view, int firstVisibleItem, int visibleItemCount, int totalItemCount) {
            if (firstVisibleItem + visibleItemCount >= totalItemCount
                    && totalItemCount != 0
                    && totalItemCount != listView.getHeaderViewsCount()
                    + listView.getFooterViewsCount()) {
                if (onLoadMoreCommand != null) {
                    methodInvoke.onNext(totalItemCount);
                }
            }
        }
    }

    public static class ListViewScrollDataWrapper {
        public int firstVisibleItem;
        public int visibleItemCount;
        public int totalItemCount;
        public int scrollState;

        public ListViewScrollDataWrapper(int scrollState, int firstVisibleItem, int visibleItemCount, int totalItemCount) {
            this.firstVisibleItem = firstVisibleItem;
            this.visibleItemCount = visibleItemCount;
            this.totalItemCount = totalItemCount;
            this.scrollState = scrollState;
        }
    }
}

File: dxbase/src/main/java/com/dxmovie/dxbase/binding/viewadapter/listview/ViewAdapter.kt
package com.dxmovie.dxbase.binding.viewadapter.listview

import android.widget.AbsListView
import android.widget.AdapterView
import android.widget.ListView
import androidx.databinding.BindingAdapter
import com.dxmovie.dxbase.binding.command.BindingCommandWithParams
import io.reactivex.subjects.PublishSubject
import java.util.concurrent.TimeUnit

class ViewAdapter {

    companion object{
        @JvmStatic
        @BindingAdapter(value = ["onScrollChangeCommand", "onScrollStateChangedCommand"], requireAll = false)
        fun onScrollChangeCommand(listView: ListView,
                                  onScrollChangeCommand: BindingCommandWithParams<ListViewScrollDataWrapper?>?,
                                  onScrollStateChangedCommand: BindingCommandWithParams<Int?>?) {
            listView.setOnScrollListener(object : AbsListView.OnScrollListener {
                private var scrollState = 0
                override fun onScrollStateChanged(view: AbsListView, scrollState: Int) {
                    this.scrollState = scrollState
                    onScrollStateChangedCommand?.executeAction(scrollState)
                }

                override fun onScroll(view: AbsListView, firstVisibleItem: Int, visibleItemCount: Int, totalItemCount: Int) {
                    onScrollChangeCommand?.executeAction(ListViewScrollDataWrapper(scrollState, firstVisibleItem, visibleItemCount, totalItemCount))
                }
            })
        }


        @JvmStatic
        @BindingAdapter(value = ["onItemClickCommand"], requireAll = false)
        fun onItemClickCommand(listView: ListView, onItemClickCommand: BindingCommandWithParams<Int?>?) {
            listView.onItemClickListener = AdapterView.OnItemClickListener { parent, view, position, id -> onItemClickCommand?.executeAction(position) }
        }


        @JvmStatic
        @BindingAdapter("onLoadMoreCommand")
        fun onLoadMoreCommand(listView: ListView, onLoadMoreCommand: BindingCommandWithParams<Int?>) {
            listView.setOnScrollListener(OnScrollListener(listView, onLoadMoreCommand))
        }

        class OnScrollListener(listView: ListView, onLoadMoreCommand: BindingCommandWithParams<Int?>) : AbsListView.OnScrollListener {
            private val methodInvoke = PublishSubject.create<Int>()
            private val onLoadMoreCommand: BindingCommandWithParams<Int?>?
            private val listView: ListView
            override fun onScrollStateChanged(view: AbsListView, scrollState: Int) {}
            override fun onScroll(view: AbsListView, firstVisibleItem: Int, visibleItemCount: Int, totalItemCount: Int) {
                if (firstVisibleItem + visibleItemCount >= totalItemCount && totalItemCount != 0 && (totalItemCount != listView.headerViewsCount
                                + listView.footerViewsCount)) {
                    if (onLoadMoreCommand != null) {
                        methodInvoke.onNext(totalItemCount)
                    }
                }
            }

            init {
                this.onLoadMoreCommand = onLoadMoreCommand
                this.listView = listView
                methodInvoke.throttleFirst(1, TimeUnit.SECONDS)
                        .subscribe { integer -> onLoadMoreCommand.executeAction(integer) }
            }
        }

        class ListViewScrollDataWrapper(var scrollState: Int, var firstVisibleItem: Int, var visibleItemCount: Int, var totalItemCount: Int)
    }
}